test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java

```diff
@@ -88,17 +89,45 @@ public void apply(PartitionUpdate update)
         }
     }
 
+    private abstract static class TruncatableVirtualTable extends AbstractVirtualTable
+    {
+        private boolean isTruncated = false;
+
+        TruncatableVirtualTable(TableMetadata metadata)
+        {
+            super(metadata);
+        }
+
+        @Override
+        public DataSet data()
+        {
+            if (isTruncated)
```

Comment:
Just a simple way to emulate truncation. Theoretically I could use a collection, smth like that:
```
TruncatableVirtualTable(TableMetadata metadata, Map<String, Integer> refToCleanupData) 

public void truncate()
{
        refToCleanupData.clear();
}
```
```
Map<String, Integer> truncatableData = <some data>;
VirtualTable vt4 = new TruncatableVirtualTable(vt4Metadata, truncatableData)
{

    public DataSet data()
    {
            SimpleDataSet data = new SimpleDataSet(metadata());
            truncatableData.forEach((key, value) -> data.row(key).column("value", value));
            return data;
    }

}
```

This way looks more similar to a real use case. Please, let me know if this one is more preferable for a unit test.

Reply:
Checking that perfoming a TRUNCATE call the `truncate` method is probably enough as the rest is dependent on the VT implementation. I would simply check that if we perform multiple TRUNCATE all the calls are transmitted to the VT and not only the first one.

Reply:
> Checking that perfoming a TRUNCATE call the truncate method is probably enough as the rest is dependent on the VT implementation. 

That's what I thought too and decided to keep the simpler version.

> I would simply check that if we perform multiple TRUNCATE all the calls are transmitted to the VT and not only the first one.

This statement sounds a bit confusing. I'm not sure where the assumption that "if the first call is transmitted to the VT, the second one may not be" comes from. In fact, it is pretty easy to substitute the flag by a counter and add the corresponding assertion, but I want to ensure it really makes sense. For me such a counting sounds like a testing of completely another logic ("routing part" is deterministic and is not really related to the VTs). Also following that assumption we may want to change other tests, for example, ensure that ``UPDATE`` is applied twice and so on.

Please, let me know your thoughts.

Reply:
I have seen case in the past where people introduced a bug in the code the tests failed to catch it. So, I was wondering if there were some scenarios that we could miss where the test could give us a false positive. I agree with you that the implementation is straight forward today but being paranoiac, I wished to make the test a bit more bullet proof. The counter approach is pretty simple and give us an extra security at low cost.